File: openclaw-tag-router/openclaw_app/adapters/audit_reason_header.py
from __future__ import annotations

import base64
import binascii
import re


AUDIT_REASON_HEADER = "X-Audit-Reason"
AUDIT_REASON_WIRE_PREFIX = "utf8-base64url-v1."
MAX_AUDIT_REASON_UTF8_BYTES = 1024
_WIRE_VALUE = re.compile(r"utf8-base64url-v1\.([A-Za-z0-9_-]+)\Z", re.ASCII)


class AuditReasonHeaderError(ValueError):
    """The audit reason header is absent or is not canonical wire data."""


def decode_audit_reason_header(value: str | None) -> str:
    if value is None:
        raise AuditReasonHeaderError("audit reason header is required")
    match = _WIRE_VALUE.fullmatch(value)
    if match is None:
        raise AuditReasonHeaderError("audit reason header uses an invalid wire format")

    payload = match.group(1)
    padding = "=" * (-len(payload) % 4)
    try:
        raw = base64.b64decode(payload + padding, altchars=b"-_", validate=True)
    except (binascii.Error, ValueError) as exc:
        raise AuditReasonHeaderError("audit reason header contains invalid base64url") from exc

    canonical_payload = base64.urlsafe_b64encode(raw).decode("ascii").rstrip("=")
    if canonical_payload != payload:
        raise AuditReasonHeaderError("audit reason header is not canonical base64url")
    if not raw:
        raise AuditReasonHeaderError("audit reason must not be empty")
    if len(raw) > MAX_AUDIT_REASON_UTF8_BYTES:
        raise AuditReasonHeaderError("audit reason exceeds 1024 UTF-8 bytes")
    try:
        return raw.decode("utf-8", errors="strict")
    except UnicodeDecodeError as exc:
        raise AuditReasonHeaderError("audit reason header is not valid UTF-8") from exc

File: openclaw-bot-center/src/media/auditReasonHeader.ts
export const AUDIT_REASON_HEADER = "X-Audit-Reason" as const;
export const AUDIT_REASON_WIRE_PREFIX = "utf8-base64url-v1." as const;

const MAX_AUDIT_REASON_UTF8_BYTES = 1024;

function bytesToBase64(bytes: Uint8Array): string {
  let binary = "";
  for (const byte of bytes) binary += String.fromCharCode(byte);
  return btoa(binary);
}

/** Encodes the exact Unicode reason into an ASCII-only versioned wire value. */
export function encodeAuditReasonHeader(reason: string): string {
  if (reason.length === 0) throw new TypeError("audit reason must not be empty");
  const bytes = new TextEncoder().encode(reason);
  if (bytes.length > MAX_AUDIT_REASON_UTF8_BYTES) {
    throw new TypeError("audit reason exceeds 1024 UTF-8 bytes");
  }
  const payload = bytesToBase64(bytes)
    .replaceAll("+", "-")
    .replaceAll("/", "_")
    .replace(/=+$/u, "");
  return AUDIT_REASON_WIRE_PREFIX + payload;
}

export function addAuditReasonHeader(
  headers: Record<string, string>,
  reason: string,
): void {
  headers[AUDIT_REASON_HEADER] = encodeAuditReasonHeader(reason);
}
